src/uct/ib/base/ib_iface.h

```diff
@@ -302,34 +302,62 @@ static inline uint8_t uct_ib_iface_get_atomic_mr_id(uct_ib_iface_t *iface)
                1; }); ++_i)
 
 /**
+ * Callback for UCT_LAZY_MEM_HANDLE mode handling in local key.
+ *
+ * It might return UCS_ERR_NO_RESOURCE if no memory in the pool.
+ */
+typedef ucs_status_t (*uct_ib_verbs_iov_callback_t)(uct_iface_h tl_iface, uint32_t length,
+                                                    void **desc_iov_p, size_t *desc_offset_p,
+                                                    uint64_t *addr_p, uint32_t *lkey_p);
+
+/**
  * Fill ibv_sge data structure by data provided in uct_iov_t
  * The function avoids copying IOVs with zero length
  *
- * @return Number of elements in sge[]
+ * if uct_iov_t::memh is UCT_LAZY_MEM_HANDLE the function copy the data into
+ * the local memory pool and uses its local key
  */
 static UCS_F_ALWAYS_INLINE
-size_t uct_ib_verbs_sge_fill_iov(struct ibv_sge *sge, const uct_iov_t *iov,
-                                 size_t iovcnt)
+ucs_status_t uct_ib_verbs_sge_fill_iov(uct_iface_h iface, void **desc_iov,
+                                       struct ibv_sge *sge,
+                                       const uct_iov_t *iov, size_t iovcnt,
+                                       size_t *sgecnt_p,
+                                       uct_ib_verbs_iov_callback_t lazy_cb)
```

Comment:
is it possible to pass memory pool instead of lazy_cb? all callbacks look pretty much the same, except rdma_read, could add the rdma_read special handling directly to the relevant function